jupyterhub/apihandlers/users.py

```diff
@@ -39,11 +40,61 @@ async def get(self):
 
 
 class UserListAPIHandler(APIHandler):
+    def _user_has_ready_spawner(self, orm_user):
+        """Return True if a user has *any* ready spawners
+
+        Used for filtering from active -> ready
+        """
+        user = self.users[orm_user]
+        return any(spawner.ready for spawner in user.spawners.values())
+
     @admin_only
     def get(self):
+        state_filter = self.get_argument("state", None)
+
+        # post_filter
+        post_filter = None
+
+        if state_filter in {"active", "ready"}:
+            # only get users with active servers
+            # an 'active' Spawner has a server record in the database
+            # which means Spawner.server != None
+            # it may still be in a pending start/stop state.
+            # join filters out users with no Spawners
+            query = (
+                self.db.query(orm.User)
+                # join filters out any Users with no Spawners
+                .join(orm.Spawner)
+                # this implicitly gets Users with *any* active server
+                .filter(orm.Spawner.server != None)
+            )
+            if state_filter == "ready":
+                # have to post-process query results because active vs ready
+                # can only be distinguished with in-memory Spawner properties
+                post_filter = self._user_has_ready_spawner
+
+        elif state_filter == "inactive":
+            # only get users with *no* active servers
+            # as opposed to users with *any inactive servers*
+            # this is the complement to the above query.
+            # how expensive is this with lots of servers?
+            query = (
+                self.db.query(orm.User)
+                .outerjoin(orm.Spawner)
+                .outerjoin(orm.Server)
+                .group_by(orm.User.id)
+                .having(func.count(orm.Server.id) == 0)
+            )
```

Comment:
The relevant syntax is [implicit line joining](https://docs.python.org/3/reference/lexical_analysis.html#implicit-line-joining) also referred to sometimes as "line continuation". This is often done with trailing backslash, which is called [explicit line joining](https://docs.python.org/3/reference/lexical_analysis.html#explicit-line-joining). Syntactically, the parentheses enclose a single [expression](https://docs.python.org/3/reference/expressions.html) (specifically, making a [parenthesized expression](https://docs.python.org/3/reference/expressions.html#parenthesized-forms)), so this is a single 'line' to the interpreter, but broken up across multiple lines in the document for readability. The parentheses serve to make it easy for the interpreter to know that it isn't done reading the expression until it gets to the closing parenthesis without needing marks on each line to tell it the expression isn't over yet.

So

```python
(
    a
    .b()
    .c()
)
```

is equivalent to [explicit line joining](https://docs.python.org/3/reference/lexical_analysis.html#explicit-line-joining) with backslash on each line before the last:

```python
a \
.b() \
.c()
```

is equivalent to

```python
a .b() .c()
```

*did you know those spaces were legal?*

is equivalent to

```python
a.b().c()
```

The reason you get concatenation with strings in this structure is that consecutive string tokens are concatenated:

```python
In [1]: "a" "b" "c"
Out[1]: 'abc'
```

I don't entirely know why Python chose to do that with strings, though.

Reply:
Thank you for this thorough and excellent answer! :heart:

> if the list contains at least one comma, it yields a tuple; otherwise, it yields the single expression that makes up the expression list.

That is the key clarification, and also explains why I see statements like `(1,)` because the single element tuple need to separate itself from being a parenthesized expression!